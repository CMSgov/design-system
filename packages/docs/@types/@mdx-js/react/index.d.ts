declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
