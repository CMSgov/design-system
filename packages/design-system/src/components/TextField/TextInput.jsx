import Mask from './Mask';
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

/**
 * <TextInput> is an internal component used by <TextField>, which wraps it and handles shared form UI like labels, error messages, etc
 * <TextInput> is also exported for advanced design system use cases, where the internal component can be leveraged to build custom form components
 * As an internal component, it's subject to more breaking changes. Exercise caution using <TextInput> outside of those special cases
 */
export function TextInput(props) {
  const {
    ariaLabel,
    errorId,
    errorMessage,
    errorPlacement,
    fieldClassName,
    inversed,
    mask,
    multiline,
    numeric,
    rows,
    size,
    setRef,
    type,
    pattern,
    ...inputProps
  } = props;

  const classes = classNames(
    'ds-c-field',
    {
      'ds-c-field--error': errorMessage,
      'ds-c-field--inverse': inversed,
    },
    mask && `ds-c-field--${mask}`,
    size && `ds-c-field--${size}`,
    fieldClassName
  );

  let inputType = type;
  if (numeric) {
    inputType = 'text';
  } else if (multiline) {
    inputType = undefined;
  }

  const ComponentType = multiline ? 'textarea' : 'input';

  /* eslint-disable react/prop-types */
  const ariaAttributes = {
    'aria-label': ariaLabel,
    // Use set `aria-invalid` based off errorMessage unless manually specified
    'aria-invalid': props['aria-invalid'] ? props['aria-invalid'] : !!errorMessage,
    // Link input to bottom placed error message
    'aria-describedby':
      errorPlacement === 'bottom' && errorMessage
        ? classNames(props['aria-describedby'], errorId)
        : undefined,
  };
  /* eslint-enable react/prop-types */

  const field = (
    <ComponentType
      {...ariaAttributes}
      className={classes}
      ref={setRef}
      rows={multiline && rows ? rows : undefined}
      inputMode={numeric ? 'numeric' : undefined}
      pattern={numeric && !pattern ? '[0-9]*' : pattern}
      type={inputType}
      {...inputProps}
    />
  );

  return mask ? <Mask mask={mask}>{field}</Mask> : field;
}

TextInput.propTypes = {
  /**
   * Apply an `aria-label` to the text field to provide additional
   * context to assistive devices.
   */
  ariaLabel: PropTypes.string,
  /**
   * Sets the initial value. Use this for an uncontrolled component; otherwise,
   * use the `value` property.
   */
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  disabled: PropTypes.bool,
  /**
   * The ID of the error message applied to the Select field.
   */
  errorId: PropTypes.string,
  errorMessage: PropTypes.node,
  /**
   * Location of the error message relative to the field input
   */
  errorPlacement: PropTypes.oneOf(['top', 'bottom']).isRequired,
  /**
   * Additional classes to be added to the input element
   */
  fieldClassName: PropTypes.string,
  /**
   * A unique ID to be used for the input field.
   */
  id: PropTypes.string.isRequired,
  /**
   * Applies the "inverse" UI theme
   */
  inversed: PropTypes.bool,
  /**
   * Apply formatting to the field that's unique to the value
   * you expect to be entered. Depending on the mask, the
   * field's appearance and functionality may be affected.
   */
  mask: PropTypes.oneOf(['currency', 'phone', 'ssn', 'zip']),
  /**
   * Whether or not the text field is a multiline text field
   */
  multiline: PropTypes.bool,
  name: PropTypes.string,
  /**
   * Sets `inputMode`, `type`, and `pattern` to improve accessiblity and consistency for number fields. Use this prop instead of `type="number"`, see [here](https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/) for more information.
   */
  numeric: PropTypes.bool,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  /**
   * @hide-prop HTML `input` [pattern](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#htmlattrdefpattern).
   */
  pattern: PropTypes.string,
  /**
   * Optionally specify the number of visible text lines for the field. Only
   * applicable if this is a multiline field.
   */
  rows: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  setRef: PropTypes.func,
  /**
   * Set the max-width of the input either to `'small'` or `'medium'`.
   */
  size: PropTypes.oneOf(['small', 'medium']),
  /**
   * HTML `input` [type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#<input>_types) attribute. If you are using `type=number` please use the numeric prop instead.
   */
  type: PropTypes.string.isRequired,
  /**
   * Sets the input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `defaultValue`.
   */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

export default TextInput;
