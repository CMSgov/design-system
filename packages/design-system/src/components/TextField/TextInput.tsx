import React, { FunctionComponent } from 'react';
import Mask from './Mask';
import classNames from 'classnames';

export type TextInputDefaultValue = string | number;
export type TextInputValue = string | number;

export interface TextInputProps {
  /**
   * Apply an `aria-label` to the text field to provide additional
   * context to assistive devices.
   */
  ariaLabel?: string;
  /**
   * Sets the initial value. Use this for an uncontrolled component; otherwise,
   * use the `value` property.
   */
  defaultValue?: TextInputDefaultValue;
  disabled?: boolean;
  /**
   * The ID of the error message applied to the Select field.
   */
  errorId?: string;
  errorMessage?: React.ReactNode;
  /**
   * Location of the error message relative to the field input
   */
  errorPlacement?: 'top' | 'bottom';
  /**
   * Additional classes to be added to the field element
   */
  fieldClassName?: string;
  /**
   * A unique `id` to be used on the text field.
   */
  id?: string;
  /**
   * Applies the "inverse" UI theme
   */
  inversed?: boolean;
  /**
   * Apply formatting to the field that's unique to the value
   * you expect to be entered. Depending on the mask, the
   * field's appearance and functionality may be affected.
   */
  mask?: 'currency' | 'phone' | 'ssn' | 'zip';
  /**
   * Whether or not the text field is a multiline text field
   */
  multiline?: boolean;
  name?: string;
  /**
   * Sets `inputMode`, `type`, and `pattern` to improve accessiblity and consistency for number fields. Use this prop instead of `type="number"`, see [here](https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/) for more information.
   */
  numeric?: boolean;
  onBlur?: (...args: any[]) => any;
  onChange?: (...args: any[]) => any;
  /**
   * @hide-prop HTML `input` [pattern](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#htmlattrdefpattern).
   */
  pattern?: string;
  /**
   * Optionally specify the number of visible text lines for the field. Only
   * applicable if this is a multiline field.
   */
  rows?: number | string;
  setRef?: (...args: any[]) => any;
  /**
   * Set the max-width of the input either to `'small'` or `'medium'`.
   */
  size?: 'small' | 'medium';
  /**
   * HTML `input` [type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#<input>_types) attribute. If you are using `type=number` please use the numeric prop instead.
   */
  type: string;
  /**
   * Sets the input's `value`. Use this in combination with `onChange`
   * for a controlled component; otherwise, set `defaultValue`.
   */
  value?: TextInputValue;
}

export type OmitProps = 'size' | 'ref';

/**
 * <TextInput> is an internal component used by <TextField>, which wraps it and handles shared form UI like labels, error messages, etc
 * <TextInput> is also exported for advanced design system use cases, where the internal component can be leveraged to build custom form components
 * As an internal component, it's subject to more breaking changes. Exercise caution using <TextInput> outside of those special cases
 */
const TextInput: FunctionComponent<
  Omit<React.ComponentPropsWithoutRef<'textarea'>, OmitProps> &
    Omit<React.ComponentPropsWithoutRef<'input'>, OmitProps> &
    TextInputProps
> = (props: TextInputProps) => {
  const {
    ariaLabel,
    errorId,
    errorMessage,
    errorPlacement,
    fieldClassName,
    inversed,
    mask,
    multiline,
    numeric,
    rows,
    size,
    setRef,
    type,
    pattern,
    ...inputProps
  } = props;

  const classes = classNames(
    'ds-c-field',
    {
      'ds-c-field--error': errorMessage,
      'ds-c-field--inverse': inversed,
    },
    mask && `ds-c-field--${mask}`,
    size && `ds-c-field--${size}`,
    fieldClassName
  );

  let inputType = type;
  if (numeric) {
    inputType = 'text';
  } else if (multiline) {
    inputType = undefined;
  }

  const ComponentType = multiline ? 'textarea' : 'input';

  /* eslint-disable react/prop-types */
  const ariaAttributes = {
    'aria-label': ariaLabel,
    // Use set `aria-invalid` based off errorMessage unless manually specified
    'aria-invalid': props['aria-invalid'] ? props['aria-invalid'] : !!errorMessage,
    // Link input to bottom placed error message
    'aria-describedby':
      errorPlacement === 'bottom' && errorMessage
        ? classNames(props['aria-describedby'], errorId)
        : undefined,
  };

  const numberRows: number = typeof rows === 'string' ? parseInt(rows) : rows;
  /* eslint-enable react/prop-types */
  const field = (
    <ComponentType
      {...ariaAttributes}
      className={classes}
      ref={setRef}
      rows={multiline && numberRows ? numberRows : undefined}
      inputMode={numeric ? 'numeric' : undefined}
      pattern={numeric && !pattern ? '[0-9]*' : pattern}
      type={inputType}
      {...inputProps}
    />
  );

  return mask ? <Mask mask={mask}>{field}</Mask> : field;
};

export default TextInput;
