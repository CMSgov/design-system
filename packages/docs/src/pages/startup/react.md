---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
