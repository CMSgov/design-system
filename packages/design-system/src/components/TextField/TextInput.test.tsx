import React from 'react';
import TextInput, { OmitProps, TextInputProps } from './TextInput';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const defaultProps: Omit<React.ComponentPropsWithRef<'textarea'>, OmitProps> &
  Omit<React.ComponentPropsWithRef<'input'>, OmitProps> &
  TextInputProps = {
  name: 'spec-field',
  inputRef: jest.fn(),
  id: 'static-id',
  type: 'text',
  errorPlacement: 'top',
};

function renderInput(props = {}) {
  return render(<TextInput {...defaultProps} {...props} />);
}

function getInput() {
  return screen.getByRole('textbox');
}

describe('TextInput', function () {
  it('is an input field', () => {
    renderInput();

    expect(getInput().tagName).toBe('INPUT');
    expect(getInput().getAttribute('rows')).toBeNull();
    expect(getInput()).toMatchSnapshot();
  });

  it('is a textarea', () => {
    renderInput({ multiline: true });

    expect(getInput().tagName).toBe('TEXTAREA');
    expect(getInput().getAttribute('rows')).toBeNull();
    expect(getInput().getAttribute('type')).toBeNull();
    expect(getInput()).toMatchSnapshot();
  });

  it('is a password field', () => {
    const { container } = renderInput({ type: 'password' });
    // The password field doesn't have an accessible role!
    const input = container.querySelector('.ds-c-field');
    expect(input.getAttribute('type')).toBe('password');
  });

  it('is disabled', () => {
    renderInput({ disabled: true });
    expect(getInput().getAttribute('disabled')).toBe('');
  });

  it('has error', () => {
    renderInput({ errorMessage: 'Error' });
    expect(getInput().getAttribute('aria-invalid')).toBe('true');
    expect(getInput().classList.contains('ds-c-field--error')).toBe(true);
  });

  it('handles bottom placed error', () => {
    renderInput({
      errorMessage: 'Error',
      errorPlacement: 'bottom',
      errorId: '1_error',
      'aria-describedby': '1_label',
    });

    expect(getInput().getAttribute('aria-invalid')).toBe('true');
    expect(getInput().getAttribute('aria-describedby')).toBe('1_label 1_error');
    expect(getInput().classList.contains('ds-c-field--error')).toBe(true);
    expect(getInput()).toMatchSnapshot();
  });

  it('has inversed theme', () => {
    renderInput({ inversed: true });
    expect(getInput().classList.contains('ds-c-field--inverse')).toBe(true);
  });

  it('has a defaultValue', () => {
    renderInput({ defaultValue: 'Yay' });
    expect(getInput().getAttribute('value')).toEqual('Yay');
  });

  it('has a value', () => {
    const value = 'Yay';
    renderInput({ value, onChange: () => {} });
    expect(getInput()).toHaveValue(value);
  });

  it('shows 5 rows of text', () => {
    renderInput({
      multiline: true,
      rows: 5,
    });
    expect(getInput().getAttribute('rows')).toBe('5');
  });

  it('adds className to field', () => {
    renderInput({ fieldClassName: 'bar' });
    expect(getInput().classList.contains('ds-c-field')).toBe(true);
    expect(getInput().classList.contains('bar')).toBe(true);
  });

  it('adds size classes to input (small)', () => {
    renderInput({ size: 'small' });
    expect(getInput().classList.contains('ds-c-field--small')).toBe(true);
  });

  it('adds size classes to input (medium)', () => {
    renderInput({ size: 'medium' });
    expect(getInput().classList.contains('ds-c-field--medium')).toBe(true);
  });

  it('adds min/max input attributes', () => {
    renderInput({ min: 1, max: 10 });
    expect(getInput().getAttribute('min')).toBe('1');
    expect(getInput().getAttribute('max')).toBe('10');
  });

  it('adds aria-label attribute', () => {
    const ariaLabel = 'Foo';
    renderInput({ ariaLabel });
    expect(getInput().getAttribute('aria-label')).toBe(ariaLabel);
  });

  it('adds aria-describedby attribute', () => {
    renderInput({ 'aria-describedby': '1_label' });
    expect(getInput().getAttribute('aria-describedby')).toBe('1_label');
  });

  it('adds undocumented prop to input field', () => {
    renderInput({ 'data-foo': 'bar' });
    expect(getInput().getAttribute('data-foo')).toBe('bar');
  });

  it('calls onBlur', () => {
    const onBlur = jest.fn();
    renderInput({ onBlur });
    fireEvent.blur(getInput());
    expect(onBlur).toHaveBeenCalled();
  });

  it('calls onChange', () => {
    const onChange = jest.fn();
    renderInput({ onChange });
    userEvent.type(getInput(), 'hello world');
    expect(onChange).toHaveBeenCalled();
  });
});
