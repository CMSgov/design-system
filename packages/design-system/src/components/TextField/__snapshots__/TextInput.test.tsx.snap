// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput handles bottom placed error 1`] = `
<input
  aria-describedby="1_label 1_error"
  aria-invalid="true"
  class="ds-c-field ds-c-field--error"
  id="static-id"
  name="spec-field"
  type="text"
/>
`;

exports[`TextInput is a textarea 1`] = `
<textarea
  aria-invalid="false"
  class="ds-c-field"
  id="static-id"
  name="spec-field"
/>
`;

exports[`TextInput is an input field 1`] = `
<input
  aria-invalid="false"
  class="ds-c-field"
  id="static-id"
  name="spec-field"
  type="text"
/>
`;
